### PR TITLE
Add tests for the vesting update procedure

### PR DESCRIPTION
**Pull request by cjjdespres**
## Explain your changes:

Tests have been added to ensure that the vesting parameter update equations in the slot reduction MIP preserve the guarantees included in the MIP, namely that the minimum balance at the hardfork slot itself is unchanged, and funds unlock at the same system times as they would have had the hardfork not occurred.

## Explain how you tested your changes:

I ran the tests manually and confirmed that they all passed.

## Checklist:

- [x] Dependency versions are unchanged
- [x] Modified the current draft of release notes with details on what is completed or incomplete within this project
   (not applicable, I think)
- [x] Document code purpose, how to use it
- [x] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? Not applicable.

### CODE REVIEW

File: src/lib/mina_base/account_timing.ml [hunk 1]
```diff
@@ -67,7 +67,7 @@ module As_record = struct
     ; vesting_period : 'slot_span
     ; vesting_increment : 'amount
     }
-  [@@deriving equal, hlist, fields, annot]
+  [@@deriving equal, hlist, fields, annot, sexp_of]
```
Comment (glyh) [inline src/lib/mina_base/account_timing.ml:70]: Could we use yojson instead of sexp?

EDIT: looks like this is enforced by Janestreet's quickcheck. I've ran into this in other parts of our code, so unfortunately, no.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+  let open Quickcheck.Generator.Let_syntax in
+  let%bind vesting_period =
+    Global_slot_span.gen_incl spec.vesting_period_min spec.vesting_period_max
+  in
+  let%bind vesting_end =
+    Global_slot_since_genesis.(
+      gen_incl (of_uint32 spec.vesting_end_min) (of_uint32 spec.vesting_end_max))
+  in
+  let%bind cliff_time =
+    Global_slot_since_genesis.(gen_incl (of_int 1) vesting_end)
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:46]: What would happen if cliff_time = vesting_end?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+      gen_incl spec.initial_minimum_balance_min spec.initial_minimum_balance_max)
+  in
+  Account.gen_vesting_details ~cliff_time ~vesting_end ~vesting_period
+    initial_minimum_balance
+
+(* Generate an account timing with a "reasonable" vesting schedule, i.e., one
+   that won't take ages to finish *)
+let gen_small_timing =
+  (* At 90s slot time, this slot will occur 375 years after slot 0 *)
+  let vesting_end_max = UInt32.(div max_int (of_int 32)) in
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:62]: Why 32, is this just a random constant to shift upper boundary to a period human care while making it proportional to `UInt32.max_int`?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+    ; vesting_end_max
+    ; initial_minimum_balance_min = Balance.one
+    ; initial_minimum_balance_max = Balance.max_int
+    }
+  in
+  gen_vesting_timing_with ~spec
+
+let gen_hardfork_timing =
+  Quickcheck.Generator.weighted_union
+    [ (1.0, gen_small_timing); (1.0, gen_large_period_timing) ]
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:95]: <del>Why not `union`?</del>

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+
+(** Return the slot at which the account completes vesting. This is the earliest
+    slot at which the current minimum balance for the account is zero. The
+    difference in balance between this returned slot and the previous slot will
+    be the [cliff_amount] if the account vests completely at its [cliff_time],
+    and otherwise will be the [vesting_increment] *)
+let final_vesting_slot (t : Account_timing.as_record) =
+  (* timing_final_vesting_slot assumes vesting_increment is non-zero, but we
+     want to be able to handle such accounts here. *)
+  if Amount.(equal t.vesting_increment zero) then t.cliff_time
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:105]: Could we move this logic inside `Account.timing_final_vesting_slot` so it deals with corner case as well?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+  else Account.timing_final_vesting_slot @@ Account.Timing.of_record t
+
+(** Generate a slot at which the account in question is vesting. *)
+let gen_slot_during_vesting ~(timing : Account_timing.as_record) =
+  let vesting_end = final_vesting_slot timing in
+  let half_period =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:111]: We will happen if the vesting period is 1here?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+    timing.vesting_period |> Global_slot_span.to_uint32
+    |> UInt32.(Fn.flip div (of_int 2))
+    |> Global_slot_span.of_uint32
+  in
+  let before_cliff =
+    let half_period_before_cliff =
+      Global_slot_since_genesis.(
+        sub timing.cliff_time half_period |> Option.value ~default:zero)
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:119]: I'm trying to dig down the case when subtraction returns a `None`. There's neither doc nor interface... What a hell of Functor and Inclusion, I'm yet to find the definition of this function.

Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:119]: EDIT: 
```ocaml
  let sub x y = if x < y then None else Some (N.sub x y)
```

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+    let slot_right_before_cliff =
+      Global_slot_since_genesis.(
+        sub timing.cliff_time Global_slot_span.one |> Option.value ~default:zero)
+    in
+    (slot_right_before_cliff, slot_right_before_cliff)
+  in
+  let at_cliff = (timing.cliff_time, timing.cliff_time) in
+  let right_after_cliff =
+    Global_slot_since_genesis.(
+      (* If cliff is too big, just exclude this case entirely *)
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml]: Is this because our migration doesn't work for absurdly long vesting schedule? If so, could you document this?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 4 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:95]: EDIT: probably we want to tweak them in the future.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+  let after_cliff =
+    let just_before_end =
+      Global_slot_since_genesis.(
+        sub vesting_end Global_slot_span.one |> Option.value ~default:zero)
+    in
+    (timing.cliff_time, just_before_end)
+  in
+  (* We want to pick slots from intervals that are non-empty and are contained
+     in the period before vesting end *)
+  let intervals =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:151]: For now, slots holding these property are carefully selected, could we generate more random slots?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+             Global_slot_since_genesis.(compare low high) <= 0
+             && Global_slot_since_genesis.(compare high vesting_end) < 0
+           then Some (weight, Global_slot_since_genesis.(gen_incl low high))
+           else None )
+  in
+  Quickcheck.Generator.weighted_union intervals
+
+(** Generate a slot at which the account has finished vesting. Note that the
+    [final_vesting_slot] is such a slot! *)
+let gen_slot_after_vesting ~(timing : Account_timing.as_record) =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml]: I'm wondering the semantics here. Last slot is technically still unlocking funds, no? If so, we shouldn't use `gen_incl` at least on the left end of the interval it seems?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+
+let gen_not_vesting_at_hardfork =
+  let open Quickcheck.Generator.Let_syntax in
+  let%bind timing = gen_hardfork_timing in
+  let%map hardfork_slot = gen_slot_after_vesting ~timing in
+  (timing, hardfork_slot)
+
+(** Generate any account timing with hardfork slot that can occur before or
+    after vesting ends *)
+let gen_any_at_hardfork =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:202]: <del>This doesn't seem to generate a slot where an account haven't start vesting?</del>

EDIT: `gen_slot_during_vesting` generate such case it seems. But I'd personally prefer an alternative name for this function as it's a bit confusing.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 12]
```diff
@@ -0,0 +1,623 @@
...
+let gen_hardfork_timing =
+  Quickcheck.Generator.weighted_union
+    [ (1.0, gen_small_timing); (1.0, gen_large_period_timing) ]
+
+(** Return the slot at which the account completes vesting. This is the earliest
+    slot at which the current minimum balance for the account is zero. The
+    difference in balance between this returned slot and the previous slot will
+    be the [cliff_amount] if the account vests completely at its [cliff_time],
+    and otherwise will be the [vesting_increment] *)
+let final_vesting_slot (t : Account_timing.as_record) =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml]: `Account_timing` and `Account.Timing` is used interchanged in this test, could we unify them into one?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 13]
```diff
@@ -0,0 +1,623 @@
...
+      ]
+  in
+  (timing, hardfork_slot)
+
+module Vesting_record = struct
+  (** A form of [Account.Timing.As_record.t] where all the fields are lifted to
+      the same [UInt64.t] type, so the arithmetic in the vesting update
+      equations becomes simpler. *)
+  type t =
+    (unit, UInt64.t, UInt64.t, UInt64.t, UInt64.t) Account.Timing.As_record.t
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml]: NOTE for myself: `'slot` and `'slot_span` are UInt32 in our real implementation.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 14]
```diff
@@ -0,0 +1,623 @@
...
+  [@@deriving equal]
+
+  (** A form of [Account.Timing.As_record.t] with all fields stored as their
+      string representations, to make [sexp_of] operations in tests easier. *)
+  type as_strings =
+    (unit, string, string, string, string) Account.Timing.As_record.t
+  [@@deriving sexp_of]
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml]: Hmm, why can't we derive `sexp_of` on `t` and remove this completely?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 15]
```diff
@@ -0,0 +1,623 @@
...
+    ; vesting_increment = t.vesting_increment |> Amount.to_uint64
+    }
+
+  (** Convert to a regular [Account_timing.as_record] by clamping all the values
+      that might be out of range, as specified in the MIP *)
+  let to_record (t : t) : Account_timing.as_record =
+    { is_timed = true
+    ; initial_minimum_balance = t.initial_minimum_balance |> Balance.of_uint64
+    ; cliff_time =
+        t.cliff_time |> clamp_uint64_to_uint32
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:248]: Shouldn't we throw an error instead of clamping?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 16]
```diff
@@ -0,0 +1,623 @@
...
+        |> Global_slot_since_genesis.of_uint32
+    ; cliff_amount = t.cliff_amount |> Amount.of_uint64
+    ; vesting_period =
+        t.vesting_period |> clamp_uint64_to_uint32 |> Global_slot_span.of_uint32
+    ; vesting_increment = t.vesting_increment |> Amount.of_uint64
+    }
+
+  (** Calculate the total number of iterations needed for this account to vest
+      completely, beyond the initial cliff unlock. This is [None] (undefined) if
+      the vesting increment is zero and the cliff amount is smaller than the
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:258]: Probably this is an `or` instead of an `and`? BTW the implementation returns `Some zero` in the later case it seems.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 17]
```diff
@@ -0,0 +1,623 @@
...
+        if equal Infix.(balance_to_unlock mod t.vesting_increment) zero then
+          (* The account unlocks an equal amount of funds during each iteration *)
+          Some full_increment_iterations
+        else
+          (* The account needs one more iteration to unlock the last little bit
+             of funds. Note: if this happens, then full_increment_iterations
+             will necessarily be well below UInt64.max_int, because division by
+             t.vesting_increment will have decreased balance_to_unlock by a
+             factor of at least two. *)
+          Some Infix.(full_increment_iterations + one))
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:285]: This doesn't match up with the definition in Slot Reduction MIP. They're rounding down with a `div`, it seems?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 18]
```diff
@@ -0,0 +1,623 @@
...
+      global_slot |> Global_slot_since_genesis.to_uint32 |> UInt64.of_uint32
+    in
+    match vesting_iterations t with
+    | None ->
+        false
+    | Some iterations ->
+        UInt64.(
+          compare global_slot t.cliff_time >= 0
+          && compare iterations
+               Infix.((global_slot - t.cliff_time) / t.vesting_period)
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:300]: I'd use multiplication as it's rounding down here. if we're at the very last slot this sub clause will return false, but there's one additional slot, so it should be true instead.

That said, the definition in Slot Reduction MIP is vague to say the least, need clarification.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 19]
```diff
@@ -0,0 +1,623 @@
...
+             > 0)
+
+  (** True if an account has not started vesting *)
+  let not_yet_vesting ~global_slot (t : t) =
+    let global_slot =
+      global_slot |> Global_slot_since_genesis.to_uint32 |> UInt64.of_uint32
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:306]: I wonder what would happen if the conversion failed, in this case and all other conversion that shrinks the number of bits a var take? Will we throw or continue silently?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 20]
```diff
@@ -0,0 +1,623 @@
...
+    in
+    UInt64.compare global_slot t.cliff_time < 0
+
+  (** True if an account is actively vesting, as defined by the slot reduction MIP *)
+  let is_actively_vesting ~global_slot (t : t) =
+    not_yet_vesting ~global_slot t || is_partially_vested ~global_slot t
+
+  (** Hardfork adjustment assuming that t is actively vesting *)
+  let actively_vesting_hardfork_adjustment ~hardfork_slot (t : t) =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:330]: Probably not very relevant, but I believe it's possible for us to have a general formula given the factor of slot reduction 2, so we could reuse these bits in the future. 

Not something that has to be done in this PR, though.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 21]
```diff
@@ -0,0 +1,623 @@
...
+        ; vesting_period =
+            (* vesting period is in the uint32 range, so this will not wrap *)
+            Infix.(of_int 2 * t.vesting_period)
+        }
+      else
+        (* t is partially but not fully vested *)
+        { t with
+          initial_minimum_balance =
+            (let balance_after_cliff =
+               assert (compare t.initial_minimum_balance t.cliff_amount > 0) ;
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:351]: <del>I believe `initial_minimum_balance` and `cliff_amount` could be equal? Unsure though. </del>

NVM, `is_partially_vested` eliminated the equal condition.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 22]
```diff
@@ -0,0 +1,623 @@
...
+                 || compare t.vesting_increment
+                      (div max_int elapsed_vesting_periods)
+                    <= 0 ) ;
+               Infix.(t.vesting_increment * elapsed_vesting_periods)
+             in
+             assert (
+               compare balance_after_cliff incremental_unlocked_balance >= 0 ) ;
+             Infix.(balance_after_cliff - incremental_unlocked_balance) )
+        ; cliff_time =
+            (* All the times and spans are in the uint32 range, so none of this
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:370]: `uint64`?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 23]
```diff
@@ -0,0 +1,623 @@
...
+    t |> of_record |> hardfork_adjustment ~hardfork_slot |> to_record)
+
+let is_actively_vesting ~global_slot (t : Account.Timing.as_record) =
+  Vesting_record.(t |> of_record |> is_actively_vesting ~global_slot)
+
+(** Test that the generator for not-actively-vesting accounts produces accounts
+    that aren't considered actively vesting *)
+let not_vesting_after_vesting () =
+  Quickcheck.test gen_not_vesting_at_hardfork ~f:(fun (timing, hardfork_slot) ->
+      [%test_pred: Global_slot_since_genesis.t]
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:402]: Is this PPX necessary?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 24]
```diff
@@ -0,0 +1,623 @@
...
+               compare balance_after_cliff incremental_unlocked_balance >= 0 ) ;
+             Infix.(balance_after_cliff - incremental_unlocked_balance) )
+        ; cliff_time =
+            (* All the times and spans are in the uint32 range, so none of this
+               will overflow *)
+            Infix.(
+              hardfork_slot
+              + of_int 2
+                * ( t.vesting_period
+                  - ((hardfork_slot - t.cliff_time) mod t.vesting_period) ))
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:376]: Okay, this is calculating the time difference of the closest vesting slot to the HF slot. Double the span and add back the HF slot to get the new slot of closest vesting. Make sense.

**Comment by cjjdespres** [comment]
There are still more test variants I could add (explicitly simulating the vesting times of accounts, for instance) but it's otherwise complete. I wanted to have this up for review and perhaps merging before I kept going.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml [hunk 25]
```diff
@@ -0,0 +1,623 @@
...
+               Global_slot_span.(
+                 compare
+                   ( Global_slot_since_genesis.diff vesting_end hardfork_slot
+                   |> Option.value_exn )
+                   one)
+               >= 0
+             then Some (timing, hardfork_slot, vesting_end)
+             else None )
+    in
+    let total_vesting_slots =
```
Comment (glyh) [inline src/lib/mina_base/test/slot_reduction_vesting_test.ml:598]: This is a span, not number of slots it seems? if it's number of slot we need to add 1.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 2 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:46]: That means the account should complete vesting at `cliff_time`. In practice, [`gen_vesting_details`](https://github.com/MinaProtocol/mina/blob/compatible/src/lib/mina_base/account.ml#L949) will set `cliff_amount` to `initial_minimum_balance`.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 3 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:62]: I should document that. Yes, I just chose some small-ish value that would result in a vesting schedule that's comfortably within the guarantees provided by the MIP.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 5 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:105]: Probably. I can move it in this PR.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 6 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:111]: `half_period` will end up being zero and we'll just generate a value equal to `cliff_time`.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 7 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:119]: Yeah, it's a safe version of subtraction that prevents underflow

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 8 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml]: I will document it better, but it's more the fact that time itself stops at the maximum slot, so if `cliff_time = max_value` then it doesn't make sense to try to generate a value in this range.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 9 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:151]: Sure, I could add cases that are more random. I wanted to start off with a good (random) spread of values in the ranges mentioned in the [notion task](https://www.notion.so/o1labs/Implement-Unit-and-Property-Tests-for-Vesting-Consistency-Across-Hard-Fork-1f2e79b1f910807cb35ccc80376d4509).

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 10 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml]: I'll document this better, but I want this to generate a slot at which the account is not actively vesting according to how the MIP defines it. That definition is basically equivalent to the current minimum balance of the account being zero at that slot, so the `final_vesting_slot` of the timing is actually something I want to include, because that slot is the first slot at which the current minimum balance is zero.

I found some of these definitions a little tricky as well because of that - vesting isn't actually implemented with a notion of "unlocking funds" anywhere. It's all about how the `timing` determines the minimum balance of an account at any particular slot. Unlocking funds is something that happens between slots, basically.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 12 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml]: I will do that

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 14 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml]: I had it this way because the `UInt64.t` used here doesn't have a `sexp_of` instance, and I was being a bit lazy.

Actually, now that I look, I see that this `UInt64.t` (which is the one used in `account.ml`) is directly from the `integers` package, but we have an `unsigned_extended` package in `mina` that looks like it provides an `Unsigned_extended.UInt64.Stable.V1.t` with `sexp`/`yojson` instances that go through strings. Interesting. Maybe I can use that one here instead.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 15 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:248]: No, it's based on how the [MIP specifies what is to be done for accounts with specific, really long vesting schedules](https://www.notion.so/o1labs/Slot-reduction-MIP-1e2e79b1f9108043bacee34c2a154009#1f2e79b1f91080f28374eee73c851c57). To be precise, there are two cases:

1. An account has a `vesting_period` that's greater than `UInt32.max_value / 2`
2. An account would next unlock funds after the hard fork at `unlock_slot`, with the hard fork occurring at `hardfork_slot` and `unlock_slot - hardfork_slot` greater than `UInt32.max_value / 2`.

In both cases it's impossible to preserve the vesting schedules exactly, because the slot time reduction is going to cut the system lifetime of the chain roughly in half, so pre-hardfork slots between `UInt32.max_value / 2` and `UInt32.max_value` don't have any counterpart in the post-hardfork timeline. But, we still have to come up with something to do for these account timings. @georgeee and I thought that clamping would be best, because `UInt32.max_value` is the closest to what value ought to be.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 16 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:258]: I think it should be `and`. The first `if` branch deals with `initial_minimum_balance <= cliff_amount` (the account vests completely at cliff time), and then the next branch will be `(initial_minimum_balance > cliff_amount) && (vesting_increment = 0)`. In that case the account is permanently stuck with a positive minimum balance equal to `initial_minimum_balance - cliff_amount` once it starts vesting actively.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 17 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:285]: You are correct. I did it this way because I'm using `UInt64.t` values and doing it exactly like it was done in the MIP (which assumes arbitrary-precision) might wrap in the `t.initial_minimum_balance - t.cliff_amount + t.vesting_increment` calculation. I was meaning to have other people (maybe @georgeee) double-check that this was equivalent.

Using `Bigint` throughout might have been wise. I only realized the vesting equations might wrap inappropriately when I was partway through implementing this.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 18 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:300]: I did it this way because the `iterations * t.vesting_period` calculation in the MIP (which assumes arbitrary precision) might overflow here. But, let me think it through out loud:

The value `(global_slot - t.cliff_time) / t.vesting_period` is the number of full vesting iterations that have elapsed from `t.cliff_time` to `global_slot`. This rounds down when `global_slot` occurs at a point in the middle of a vesting iteration, which makes sense, since we're only partway through a vesting iteration if that happens. The `iterations` value is the number of full iterations that need to have elapsed since `cliff_time` for vesting to be finished. So, `iterations > ((global_slot - cliff_time) / vesting_period)` should be true exactly when we haven't completed enough iterations to have finished vesting.

Okay, it feels like what I've written in the code is correct.

Actually, when you said "last slot" did you mean the last slot at which the account unlocks funds? An account isn't considered actively vesting at that slot. (I mentioned that in an earlier comment, but wanted to say that here just for completeness. Sorry for the repetition).

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 19 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:306]: This particular conversion is `UInt32 -> UInt64`, so it should be lossless

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 20 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:330]: Agreed

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 22 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:370]: No, I meant that all the times were originally `UInt32` values, so they're in the corresponding range of `UInt64.t` values.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 25 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:598]: It's the number of slots, but the account isn't considered actively vesting at `vesting_end`, so this should be the length of the interval `[hardfork_slot, vesting_end)`.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 19 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:306]: Indeed. And also `Global_slot_since_genesis` is just u32 with a wrapper.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 18 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:300]: Yeah, looks like misunderstanding.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 6 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:111]: If so, isn't `before_cliff` not before cliff?

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 5 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:105]: Not crucial, though. we could do it in subsequent PRs if needed.

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 16 above)
Comment (glyh) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:258]: Indeed, my bad

File: src/lib/mina_base/test/slot_reduction_vesting_test.ml (hunk 5 above)
Comment (cjjdespres) [reply src/lib/mina_base/test/slot_reduction_vesting_test.ml:105]: Sure, might be better not to complicate this one

**Comment by cjjdespres** [comment]
!ci-build-me

**Comment by cjjdespres** [comment]
I think that should be everything, other than a few items for follow-up PRs.

**Review by glyh — APPROVED** [review]

**Comment by georgeee** [comment]
!ci-bypass-changelog

**Comment by georgeee** [comment]
Merging this because there is approval by Yihang.

Also, after checking the code briefly (not all of the code I reviewed, hence no approval), I think it's worth adding some unit tests with iteration of every slot in a short range. I.e. something limited and dumb, in addition to excellent property-based coverage.

_P.S. We need to update codeowners_
